src/base-app.js

```diff
@@ -0,0 +1,78 @@
+import {withMiddleware} from './with-middleware';
+import {withDependencies} from './with-dependencies';
+
+class FusionApp {
+  constructor() {
+    this.registered = new Map();
+    this.plugins = [];
+  }
+  register(token, Plugin) {
+    if (Plugin === undefined) {
+      Plugin = token;
+    }
+    this.plugins.push(token);
+    this.registered.set(token, Plugin);
+  }
+  configure(token, value) {
+    this.registered.set(token, value);
+  }
+  middleware(deps, middleware) {
+    if (middleware === undefined) {
+      middleware = deps;
+      this.register(withMiddleware(middleware));
+    } else {
+      this.register(
+        withDependencies(deps)(d => {
+          return withMiddleware(middleware(d));
+        })
+      );
+    }
+  }
+  resolve() {
+    this.register(this.renderer);
+    const resolved = new Map();
+    const resolving = new Set();
+    const registered = this.registered;
+    const resolvedPlugins = [];
+    // TODO: maybe could turn this into a map
+    this.plugins.forEach(function resolveToken(token) {
+      // if we have already resolved the type, return it
+      if (resolved.has(token)) {
+        return resolved.get(token);
+      }
+      // if currently resolving the same type, we have a circular dependency
+      if (resolving.has(token)) {
+        throw new Error(
+          `Cannot resolve circular dependency: ${token.toString()}`
+        );
+      }
+      // the type was never registered, throw error
+      if (!registered.has(token)) {
```

Comment:
What is the rationale for not hard failing here?

Reply:
Supporting optional dependencies with defaults. Basically I have this `fusion-tokens` repo which has two helper exports and then a bunch of agreed upon tokens. The helpers look like this:

```
export const createToken = name => () => {
  throw new Error(`Missing required value for token: ${name}.`);
};
export const createOptionalToken = (name, defaultValue) => defaultValue;
```

This pattern allows us to enforce that if you want to define an optional parameter, you **must** provide a default value for it. Additionally, if you have a required parameter, you get nice error messages.

Reply:
Are there tests for `createOptionalToken`?